receiver.go

```diff
@@ -108,39 +52,113 @@ func (r *Receiver) Receive(ctx context.Context) (*Message, error) {
 		}
 	}
 
+	// for ModeFirst, auto-accept the message
+	modeFirstAccept := func(ctx context.Context, msg *Message) (*Message, error) {
+		if r.manualCreditor != nil || receiverSettleModeValue(r.link.ReceiverSettleMode) == ModeSecond {
+			return msg, nil
+		}
+		if err := r.AcceptMessage(ctx, msg); err != nil {
+			return nil, err
+		}
+		return msg, nil
+	}
+
 	// non-blocking receive to ensure buffered messages are
 	// delivered regardless of whether the link has been closed.
 	select {
 	case msg := <-r.link.Messages:
-		// we remove the message from unsettled map as soon as it's popped off the channel
-		// This makes the unsettled count the same as messages buffer count
-		// and keeps the behavior the same as before the unsettled messages tracking was introduced
-		defer r.link.DeleteUnsettled(&msg)
 		debug(3, "Receive() non blocking %d", msg.deliveryID)
-		msg.receiver = r
-		return &msg, nil
+		msg.link = r.link
+		return modeFirstAccept(ctx, &msg)
 	case <-ctx.Done():
 		return nil, ctx.Err()
 	default:
+		// done draining messages
 	}
 
 	// wait for the next message
 	select {
 	case msg := <-r.link.Messages:
-		// we remove the message from unsettled map as soon as it's popped off the channel
-		// This makes the unsettled count the same as messages buffer count
-		// and keeps the behavior the same as before the unsettled messages tracking was introduced
-		defer r.link.DeleteUnsettled(&msg)
 		debug(3, "Receive() blocking %d", msg.deliveryID)
-		msg.receiver = r
-		return &msg, nil
+		msg.link = r.link
+		return modeFirstAccept(ctx, &msg)
+	case <-r.link.close:
+		return nil, r.link.err
 	case <-r.link.Detached:
 		return nil, r.link.err
 	case <-ctx.Done():
 		return nil, ctx.Err()
 	}
 }
 
+// Accept notifies the server that the message has been
+// accepted and does not require redelivery.
+func (r *Receiver) AcceptMessage(ctx context.Context, msg *Message) error {
+	if !msg.shouldSendDisposition() {
+		return nil
+	}
+	if err := r.messageDisposition(ctx, msg, &encoding.StateAccepted{}); err != nil {
+		return err
+	}
+	msg.settled = true
```

Comment:
This would be useful as a public field, if you're willing. Or maybe a public function.

Reply:
In hindsight I shouldn't have done this as this really isn't part of Message public surface area.  It was a simple way to prevent disposing a message multiple times which doesn't appear to be necessary.

Reply:
If it does become problematic we can address it separately.